### PR TITLE
feat/API: add publicly available `/api/v1/chat/completions` REST endpoint

### PR DESCRIPTION
**Pull request by olafurpg**
Fixes CODY-3081

This is the first PR for the project [⏭️ Launch Cody API Experimental](https://linear.app/sourcegraph/project/launch-cody-api-experimental-8fd5ec338bf4), which falls under the umbrella of moving Cody's brains to the cloud.

Previously, there was no publicly available REST API for our customers to interact with Cody.  This is a frequently requested feature from customers and prospects.

This PR adds a new `POST /api/v1/chat/completions` endpoint, which should be compatible with existing OpenAI clients. The OpenAI API format is increasingly becoming an industry standard so this seems like a good first step towards exposing a stable publicly facing API for our customers.

The goal is to add more Cody-specific APIs in the coming weeks to send chat messages and reference context.

## Test plan

See added test cases.

## Changelog

* API: new publicly available `/api/v1/chat/completions` REST endpoint
that is compatible with OpenAI clients with some restrictions. The detailed list of restrictions will eventually be documented on sourcegraph.com/docs.

### CODE REVIEW

**Comment by olafurpg** [comment]
Opened the PR to record a Loom. I will do a self-review later today before tagging for review.

**Comment by olafurpg** [comment]
Also, for the record, this is not feature complete yet, I need to test this way more thoroughly with different LLM providers. I was rushing to open the PR just to wrap up the Krakow Hackathon.

File: cmd/frontend/internal/publicrestapi/openai_handlers.go [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+
+	"github.com/sourcegraph/sourcegraph/lib/errors"
+)
+
+// serveOpenAIChatCompletions is a handler for the OpenAI /v1/chat/completions endpoint.
+func serveOpenAIChatCompletions(logger sglog.Logger, apiHandler http.Handler) func(w http.ResponseWriter, r *http.Request) (err error) {
+	return func(w http.ResponseWriter, r *http.Request) (err error) {
+		// Parse OpenAI request
+
+		logger.Info("Received OpenAI chat completion request",
```
Comment (mmanela) [inline cmd/frontend/internal/publicrestapi/openai_handlers.go:21]: I suggest not using the name OpenAI in the files or route path. I think even though we are using a protocol they defined, it is not tied to them. This can be misleading (to think its specific to openai , the company's, products)

Comment (olafurpg) [reply cmd/frontend/internal/publicrestapi/openai_handlers.go:21]: This logging statement will get removed and we should not refer to these as "OpenAI chat completion requests".

I'm fine replacing `/api/openai/v1/chat/completions` with just `/api/v1/chat/completions` because we will never have a competing implementation of `/chat/completions`. The main downside of going this route is we should ideally also follow OpenAI strictly for other endpoints like `/api/v1/models` where it might be tempting to customize the JSON object shape for our custom types (and we can still do that just under a different REST endpoint).

Comment (mmanela) [reply cmd/frontend/internal/publicrestapi/openai_handlers.go:21]: I agree but putting the name in the path doesn't solve that. We can add schema validation test to ensure we follow their schema

Comment (olafurpg) [reply cmd/frontend/internal/publicrestapi/openai_handlers.go:21]: Settled on naming the endpoint `/api/openaicompat/v1/chat/completions` to empasize the OpenAI-compatibility (not OpenAI itself)

Comment (slimsag) [reply cmd/frontend/internal/publicrestapi/openai_handlers.go:21]: my 2c: "openaicompatible" is referred to throughout the client and server codebase, and site config, to mean 'self hosted models' - so I think having this _also_ named openaicompat will be confusing for [customers, support, etc.] talking about this

Reserving `/.api/v1` for this would also not be good as I imagine we might one day want to version other parts of our API under that.

I'd go for something like `/.api/cody/v1/chat/completions`, `/.api/cody/v1/models`, etc. personally

Comment (olafurpg) [reply cmd/frontend/internal/publicrestapi/openai_handlers.go:21]: I'm starting to lean towards `/api/v1/chat/completions` and if we ever need to diverge from OpenAI shapes then we just come up with some fresh names.

> /.api/cody/

These APIs will btw not get scoped under "Cody", they will just be toplevel endpoints in the "Sourcegraph API".

> /.api

FWIW, the plan is to use `/api` instead of `/.api` to indicate that these are publicly-facing REST endpoints that can be used by customers, which also means we will need to document them to a high quality standard and provide strong backwards/forwards compatibility guarantees.

**Review by olafurpg — COMMENTED** [review]
I was so unhappy with the code/tests that I'm working on a big rewrite that should make the structure nicer. My goal is to establish good practices from the start so that we can easily scale up this API to hundreds/thousands of endpoints. One of the big problems with the current draft is that it takes 6s to link the module because we depend on our httpapi (which is huge). The changes I have locally fix that problem

**Comment by olafurpg** [comment]
OK, I've made significant changes to the implementation since I created the PR:

* New HTTP record/replay testing infrastructure. I was desperately missing this after having gotten used to this in the Cody repo. This Go implementation addresses many of the issues I've faced with Polly.js and is hopefully not going to suffer from the same XXL recording files that have endless git rebase conflicts.
* The endpoint is now `/api/v1/chat/completions`. I just removed the "Open AI" mention. Open to changing my opinion on this.
* The publicrestapi module no longer imports from the monolithic internal/httpapi or graphqlbackend modules so the tests link much faster (from 6s to <<1s) making it much more enjoyable to iterate on the code.

I will do another round of self-review before tagging people for review.

File: cmd/frontend/internal/handlerutil/handler.go [hunk 2]
```diff
@@ -13,6 +13,9 @@ import (
 	"github.com/sourcegraph/sourcegraph/lib/errors"
 )
 
+type HandlerWithErrorReturnFunc func(http.ResponseWriter, *http.Request) error
```
Comment (olafurpg) [inline cmd/frontend/internal/handlerutil/handler.go:16]: This diff is unrelated to the latest version of this PR but I'm keeping it because it took me a while to understand the relationship between these function signatures and where they are used (example: `JsonMiddleware`). Happy to remove

**Comment by slimsag** [comment]
> The goal is to add more Cody-specific APIs in the coming weeks to send chat messages and reference context.

Can you clarify this? Do you intend to stick relatively strictly to the OpenAI API alleged-spec, or do you intend to diverge from it?

File: cmd/frontend/internal/publicrestapi/chat_completions_handler.go [hunk 3]
```diff
@@ -0,0 +1,239 @@
...
+	"github.com/sourcegraph/sourcegraph/lib/errors"
+
+	completions "github.com/sourcegraph/sourcegraph/internal/completions/types"
+)
+
+// chatCompletionsHandler implements the REST endpoint /chat/completions
+type chatCompletionsHandler struct {
+	logger sglog.Logger
+
+	// apiHandler is the underlying implemenation of the Sourcegraph /.api/completions/stream endpoint.
```
Comment (slimsag) [inline cmd/frontend/internal/publicrestapi/chat_completions_handler.go:25]:
```suggestion
	// apiHandler is the underlying implementation of the Sourcegraph /.api/completions/stream endpoint.
```

Also, `/.api` or `/api`?

File: cmd/frontend/internal/publicrestapi/chat_completions_handler.go [hunk 4]
```diff
@@ -0,0 +1,239 @@
...
+	apiHandler http.Handler
+}
+
+var _ http.Handler = (*chatCompletionsHandler)(nil)
+
+func (h *chatCompletionsHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	var chatCompletionRequest CreateChatCompletionRequest
+	body, err := io.ReadAll(r.Body)
+	if err != nil {
+		http.Error(w, "io.ReadAll(r.Body) failed", http.StatusInternalServerError)
```
Comment (slimsag) [inline cmd/frontend/internal/publicrestapi/chat_completions_handler.go:39]: you are discarding the actual `err` string here which is not good; better would be to use `fmt.Sprintf` or such like `fmt.Sprintf("io.ReadAll: %s", err.Error())`

File: cmd/frontend/internal/publicrestapi/chat_completions_handler.go [hunk 5]
```diff
@@ -0,0 +1,239 @@
...
+		return
+	}
+
+	decoder := json.NewDecoder(io.NopCloser(bytes.NewBuffer(body)))
+
+	if err := decoder.Decode(&chatCompletionRequest); err != nil {
+		http.Error(w, "decoder.Decode(body) failed "+err.Error(), http.StatusInternalServerError)
```
Comment (slimsag) [inline cmd/frontend/internal/publicrestapi/chat_completions_handler.go:46]: In general avoid string concatenation in Go, better to use `fmt.Sprintf` or https://pkg.go.dev/strings#Builder for more complex cases. Concatenation is slower.

File: cmd/frontend/internal/publicrestapi/openai_handlers.go (hunk 1 above)
Comment (slimsag) [reply cmd/frontend/internal/publicrestapi/openai_handlers.go]: > FWIW, the plan is to use /api instead of /.api to indicate that these are publicly-facing REST endpoints

Not opposed to this, but you are diverging from:

* The publicly available Search REST API https://sourcegraph.com/docs/api/stream_api
* The publicly available `/.api/graphql` API which is advertised in various product pages (e.g. when you create an access token) and https://sourcegraph.com/docs/api/graphql

My advice would be don't do this to avoid creating confusion, just stick to `/.api`

File: cmd/frontend/internal/publicrestapi/chat_completions_handler.go [hunk 6]
```diff
@@ -0,0 +1,239 @@
...
+	if err = json.NewEncoder(w).Encode(chatCompletionResponse); err != nil {
+		h.logger.Error("writing /chat/completions response body", log.Error(err))
+	}
+
+}
+
+func validateChatCompletionRequest(chatCompletionRequest CreateChatCompletionRequest) string {
+
+	if chatCompletionRequest.N != nil && *chatCompletionRequest.N != 1 {
+		return "n is not supported"
```
Comment (slimsag) [inline cmd/frontend/internal/publicrestapi/chat_completions_handler.go:75]:
```suggestion
		return "n must be nil or 1"
```

File: cmd/frontend/internal/publicrestapi/chat_completions_handler.go [hunk 7]
```diff
@@ -0,0 +1,239 @@
...
+		return nil, errors.Newf("failed to unmarshal response body %s: %v", string(responseBytes), err)
+	}
+
+	return &sgResp, nil
+}
+
+func transformToOpenAIResponse(sgResp *completions.CompletionResponse, openAIReq CreateChatCompletionRequest) CreateChatCompletionResponse {
+	// Transform Sourcegraph response to OpenAI format
+	return CreateChatCompletionResponse{
+		ID:      "chat-" + generateUUID(), // You'll need to implement generateUUID
```
Comment (slimsag) [inline cmd/frontend/internal/publicrestapi/chat_completions_handler.go:210]: who is 'you' in this comment and below? are these supposed to be TODOs?

**Comment by olafurpg** [comment]
> Can you clarify this? Do you intend to stick relatively strictly to the OpenAI API alleged-spec, or do you intend to diverge from it?

The goal is to only stick with OpenAI for `/chat/completions`, `/models` and `/models/{model}` to begin with, and have other sibling endpoints that are Sourcegraph specific. If we add more functionality like multi-modal (uploading images/videos) or structured output then we should stick with OpenAI API conventions (and I suspect a good chunk of the industry will do the same).

The initial goal was to group those three endpoints under `/api/openai/v1` to build a clearer separation but I went with the new unified design (everything under `/api/v1`) because I want to move the focus away from OpenAI and towards "Sourcegraph API". We can document on sourcegraph.com/docs what subset of endpoints from the OpenAI API are supported.

File: cmd/frontend/internal/publicrestapi/chat_completions_handler.go (hunk 3 above)
Comment (olafurpg) [reply cmd/frontend/internal/publicrestapi/chat_completions_handler.go:25]: Fixed the typo. The underlying API is `/.api/completions/stream` while this endpoint is intentionally `/api` without the leading period `.` to emphasize that this is a public and customer-facing REST endpoint. I am planning to document `/api` endpoints on sourcegraph.com/docs.

File: cmd/frontend/internal/publicrestapi/golly-recordings/chat_completions.recording.yaml [hunk 8]
```diff
@@ -0,0 +1,159 @@
+recordings:
```
Comment (slimsag) [inline cmd/frontend/internal/publicrestapi/golly-recordings/chat_completions.recording.yaml:1]: I personally am not a fan of this type of request-recording-testing, because it tends to capture request variables that are truly dynamic and change every time you rerun the recording logic - and it lacks comments / indications of what parts of the request are actually important / not expected to change. I think it leads to worse quality tests and makes it easy to 'accept updated recordings' which leads to regressions.

That said, I understand others like them so no objection.

**Review by slimsag — APPROVED** [review]

**Review by olafurpg — COMMENTED** [review]
Thank you @slimsag!

File: cmd/frontend/internal/publicrestapi/golly-recordings/chat_completions.recording.yaml (hunk 8 above)
Comment (olafurpg) [reply cmd/frontend/internal/publicrestapi/golly-recordings/chat_completions.recording.yaml:1]: That's a fair concern, and I don't think you are alone with that opinion. I've experienced these downsides you mention first-hand in the Cody repo (where I also introduced them). Dealing with git conflicts in big-traffic repos is another common issue (although I suspect it won't be as big a problem here).

For me, the benefits heavily outweigh the downsides, and I am hoping to avoid repeating some of the mistakes in the Cody repo by having smaller and more narrowly-focused recording files. As soon as I had this infrastructure in place, I felt like 10x more productive hacking on the backend code and it made me feel a lot more confident that my code actually works.

Comment (olafurpg) [reply cmd/frontend/internal/publicrestapi/golly-recordings/chat_completions.recording.yaml:1]: > it tends to capture request variables that are truly dynamic

I already hit on this problem in HTTP responses in this PR and addressed it by only keeping `Content-Header` (and you can override this logic). I also made some quality-of-live improvements like encoding the request body as multi-line YAML instead of single-line JSON for improved readability.

Comment (slimsag) [reply cmd/frontend/internal/publicrestapi/golly-recordings/chat_completions.recording.yaml:1]: FWIW, this is why I encourage autogold inline testing because it lets you be more selective about which parts you want to snapshot-test, and lets you add comments explaining why those parts are particularly interesting to confirm they have not changed.

But yeah not opposed to this, just offering feedback.

Comment (olafurpg) [reply cmd/frontend/internal/publicrestapi/golly-recordings/chat_completions.recording.yaml:1]: I love autogold ❤️ thank you for building it. I am using autogold to assert the most relevant parts of the responses (like 200 status) to avoid blind LGTMs of critical regressions.

File: internal/golly/testing_credentials.go [hunk 9]
```diff
@@ -0,0 +1,6 @@
+package golly
+
+type TestingCredentials struct {
```
Comment (olafurpg) [inline internal/golly/testing_credentials.go:3]: This is in a separate file because it used to be a more advanced struct with helper methods to deal with redaction of access tokens but I removed it because it's not needed in this initial revision.

Comment (olafurpg) [reply internal/golly/testing_credentials.go:3]: aha, the redaction is not needed because we're not hashing the HTTP headers to identify requests. It's fine for now because we only use one access token per recording, but we may need to add this later.

File: cmd/frontend/internal/publicrestapi/chat_completions_handler.go [hunk 10]
```diff
@@ -0,0 +1,236 @@
...
+	req, err := http.NewRequestWithContext(r.Context(), "POST",
+		"/.api/completions/stream?api-version=1&client-name=openai-rest-api&client-version=6.0.0",
+		bytes.NewBuffer(reqBody))
+
+	if err != nil {
+		return nil, errors.Newf("failed to create request: %v", err)
+	}
+
+	// Set headers from the original request
+	for headerName, values := range r.Header {
```
Comment (chrsmith) [inline cmd/frontend/internal/publicrestapi/chat_completions_handler.go:174]: This seems a little dubious... why is this require?

Is it that the HTTP request sent to _this_ handler only accepts credentials in a different format than that of the completions API?

Either way, it seems like bad juju to ever be touching the Authorization header. I kinda assumed that the current `r.Context()` would contain the credentials/authorization status based on the _calling user_. And so when we create the faux HTTP request to the completions endpoint, it would "know" that the request is already authenticated... is that not the case?

Or am I just super-confused, and that this endpoint does not use _Sourcegraph credentials_, but instead requires an OpenAPI access key or something?

File: cmd/frontend/internal/publicrestapi/chat_completions_models.go [hunk 11]
```diff
@@ -0,0 +1,184 @@
+package publicrestapi
+
+import (
+	"encoding/json"
+
+	"github.com/sourcegraph/sourcegraph/lib/errors"
+)
+
+type CreateChatCompletionRequest struct {
```
Comment (chrsmith) [inline cmd/frontend/internal/publicrestapi/chat_completions_models.go:9]: What is this type exactly? Is this the Sourcegraph-defined API request type that our new HTTP endpoint expects to receive?

And... we expect this or perhaps just the response to match the OpenAI API response format?

IMHO some doc comments would be helpful on these type definitions to explain the what and why.

File: cmd/frontend/internal/publicrestapi/chat_completions_models.go [hunk 12]
```diff
@@ -0,0 +1,184 @@
...
+	Messages         []ChatCompletionRequestMessage          `json:"messages"`
+	Model            string                                  `json:"model"`
```
Comment (chrsmith) [inline cmd/frontend/internal/publicrestapi/chat_completions_models.go:11]: We should be super-duper clear about what is expected to go here. As there are a growing number of formats that can be used.

Is this a `model-name`, or `provider/model`? Will a `provider::api-version::model` mRef be accepted?

If this is just wrapping the completions API, then all three would be supported. But the hope is that over time we'll only have clients sending the more precise mRef format. (But perhaps that's not what we want for this particular API?)

Regardless, I think we should have a doc comment explaining what values are acceptable for this field.

File: cmd/frontend/internal/publicrestapi/chat_completions_models.go [hunk 13]
```diff
@@ -0,0 +1,184 @@
...
+type ChatCompletionRequestContentProperty struct {
+	Parts []ChatCompletionRequestMessageContentPartText `json:"parts,omitempty"`
+}
+
+type ChatCompletionRequestMessageContentPartText struct {
+	Type string `json:"type" enum:"text"`
+	Text string `json:"text"`
+}
+
+func (c *ChatCompletionRequestContentProperty) UnmarshalJSON(data []byte) error {
```
Comment (chrsmith) [inline cmd/frontend/internal/publicrestapi/chat_completions_models.go:82]: It's unfortunate that we need to write these custom JSON deserialization routines. I'm guessing it is because this is the OpenAI-specific request format we are trying to maintain compatibility with?

If so, would it be preferable to just use some OpenAI-maintained API client library instead? Which might make it easier to keep these types and associated logic up-to-date?

i.e. if we are trying to introduce an invariant that we are compatible with the  OpenAI request/response format, then we would benefit from leaning into that. Rather than just maintaining our own parallel implementation/definition?

File: cmd/frontend/internal/publicrestapi/golly-recordings/chat_completions.recording.yaml [hunk 14]
```diff
@@ -0,0 +1,159 @@
...
+  response:
+    status_code: 200
+    body: '{"completion":"yes","stopReason":"end_turn"}'
+    headers:
+    - key: Content-Type
+      value: text/plain; charset=utf-8
+- hash: d6d9aaaf5bd9242c2f8d25ead69a79495b00a944635231ad00ff78ceaaa4988f
+  request:
+    recording_date: "2024-08-07T17:00:31+02:00"
+    url: https://sourcegraph.com/.api/modelconfig/supported-models.json
```
Comment (chrsmith) [inline cmd/frontend/internal/publicrestapi/golly-recordings/chat_completions.recording.yaml:31]: Is this just for providing a static HTTP response for testing? Or are we trying to verify that the result of `/supported-models.json` will _continue_ to look like this?

Because unless our test cases are defining every model/provider explicitly, this is definitely going to be flakey as new models get introduced.

File: cmd/frontend/internal/publicrestapi/utils_test.go [hunk 15]
```diff
@@ -0,0 +1,131 @@
...
+	"github.com/stretchr/testify/assert"
+
+	"github.com/sourcegraph/sourcegraph/internal/golly"
+	"github.com/sourcegraph/sourcegraph/internal/httpcli"
+	"github.com/sourcegraph/sourcegraph/internal/modelconfig/types"
+	"github.com/sourcegraph/sourcegraph/internal/txemail"
+)
+
+func init() {
+	txemail.DisableSilently()
```
Comment (chrsmith) [inline cmd/frontend/internal/publicrestapi/utils_test.go:22]: Why is this needed? I wouldn't expect using an API to need to interact with the transactional email package? (Or perhaps `txemail` means something else.) Perhaps add a short comment as to why this is required?

File: cmd/frontend/internal/publicrestapi/utils_test.go [hunk 16]
```diff
@@ -0,0 +1,131 @@
...
+	t           *testing.T
+	Handler     http.Handler
+	Golly       *golly.Golly
+	Credentials *golly.TestingCredentials
+	HttpClient  http.Handler
+}
+
+func newTest(t *testing.T, name string) *publicrestTest {
+	MockUUID = "mocked-publicrestapi-uuid"
+	gollyDoer := golly.NewGollyDoer(t, name, httpcli.TestExternalClient)
```
Comment (chrsmith) [inline cmd/frontend/internal/publicrestapi/utils_test.go:35]: JFYI, you can get the name of the current model via `t.Name()`. (So you can remove the `name` parameter if you want to.)

And instead, just wrote things like:

```go
...
t.Run("VerifyDodads", func(t *testing.T) {
    testObj := newTest(t) // t.Name() will be "ParentTest/ParentTest2/VerifyDodads"
})

**Review by chrsmith — COMMENTED** [review]
I get the gist of what you are trying to do here, and it's kinda elegant to just wrap and simulate invoking the completions API handler... but I'm not quite sure why this is all necessary.

What precisely are we trying to do with the new `/chat/completions` API that differs from the current completions API?

Is it just wrapping the "request format" and "response format" so that they more closely match the OpenAI format? If so, is that going to broadly work?

For example, today if you use the existing chat completions API with different models from different providers, the same HTTP request will fail because of minor differences in what values are accepted in the "messages.speaker". Because under the hood, we end up converting our `CodyCompletionsProviderRequest` into any number of LLM API Provider-specific request shapes... and I don't know how confident we are that we've found a way to express the "input request" that will reliably work when used through various models.

Or, thinking aloud, would we want to have some integration test suite that verified we could send the same request to this API but using every model available to Cody Pro users, etc.?

File: cmd/frontend/internal/publicrestapi/golly-recordings/chat_completions.recording.yaml (hunk 14 above)
Comment (olafurpg) [reply cmd/frontend/internal/publicrestapi/golly-recordings/chat_completions.recording.yaml:31]: Good call! After reading the missing spec for modelconfig it became clear to me how I can avoid this request. I will remove it.

I did it this way because I was trying to test assumptions like "every chat-capable model on dotcom should be able to go through this endpoint".

Comment (olafurpg) [reply cmd/frontend/internal/publicrestapi/golly-recordings/chat_completions.recording.yaml]: I took a stab at removing this request from the tests and landed on the conclusion that it's cleaner to keep it in the recording. At the end of the day, it's the same solution whether we hardcode the modelconfig in the recording file or in a Go struct. The recording is easier to update in the future (you just remove recording and re-record) while refactoring the Go struct is more error prone (it's easy to write non-sense model configuration that doesn't reflect production usage).

File: cmd/frontend/internal/publicrestapi/utils_test.go (hunk 16 above)
Comment (olafurpg) [reply cmd/frontend/internal/publicrestapi/utils_test.go]: TIL I changed the logic to use `t.Name` for the recording name. I may add the option to override the name in the future for cases where I want more control, but I don't need this functionality now.

**Review by olafurpg — COMMENTED** [review]
Thank you @chrsmith for the review! I am going to merge this PR to post a project update today and I am happy to address any post-merge review comments with followup PRs 🙏🏻

File: cmd/frontend/internal/publicrestapi/utils_test.go (hunk 15 above)
Comment (olafurpg) [reply cmd/frontend/internal/publicrestapi/utils_test.go:22]: Not needed anymore, removed. The original implementation of this PR had to mock a lot of internal services and I copy-pasted this into the tests alongside a bunch of other non-trivial mocking logic from other files. The new implementation is a lot simpler to reason about (we're just massaging HTTP request/responses)

**Comment by olafurpg** [comment]
I am going to merge this PR even if it's still imperfect. I've opened a few followup issues:

- [CODY-3194: Fix bug where OpenAI models return empty text](https://linear.app/sourcegraph/issue/CODY-3194/fix-bug-where-openai-models-return-empty-text)
- [CODY-3210: Implement streaming `/chat/completions`](https://linear.app/sourcegraph/issue/CODY-3210/implement-streaming-chatcompletions)
- [CODY-3211: Implement cancelable `/chat/completions`](https://linear.app/sourcegraph/issue/CODY-3211/implement-cancelable-chatcompletions) (although I think this is already supported, but I want to confirm it 100%).

File: cmd/frontend/internal/publicrestapi/chat_completions_handler.go (hunk 10 above)
Comment (chrsmith) [reply cmd/frontend/internal/publicrestapi/chat_completions_handler.go:215]: Yes, that clarifies things, things.

And perhaps this is something I missed and might be useful to call out in a BIG ALL CAPS comment on the function. But the intention is to be "so compatible with OpenAI", that we would expect this endpoint to be a replacement for other tools that talk directly to OpenAI. Right?

So we didn't make this take the same API request as OpenAI as a "nice-to-have" or way to just not think about our API design. But really, it's like, for reals needing to be a drop-in replacement? (And just using Sourcegraph and its machinery to route the request to OpenAI, rather than a Sourcegraph user needing to have their own OpenAI API key, etc.)

Is that the right conceptual model?

File: cmd/frontend/internal/publicrestapi/chat_completions_models.go (hunk 12 above)
Comment (chrsmith) [reply cmd/frontend/internal/publicrestapi/chat_completions_models.go:20]: > prevents users from specifying fuzzy models something like "sonnet-3" to match the latest version. However, we can make the model semantics more flexible in the future (for example, allow prefix matching on the modelref).

Yeah, I definitely wouldn't be opposed to that. And obviously that makes this all easier to use. (Since then our API would _actually_ be compatible with OpenAI, taking `model: "gpt-4o"` rather than needing our own specific model format.

We already have most of the logic necessary in `get_model.go` and how we try to find a match between what the user requested and the models we know about. So I think we'd just need to clean that up and clarify the exact lookup rules.

e.g. if the user  `model: "gpt-4o"`, they are probably referring to the `ModelName` (what the API expects). But if they specify `openai/gpt-4o`, then I would probably expect that to be referring to the `ModelID` (which could map to `ModelName` `gpt-4o-20240808` or something.)

But yes, that's a good idea and something we should probably do at some point.

> I had to change the modelconfig.ResetMocks() function to accept custom ModelConfiguration.

Yeah, that sounds right. You'd need to provide the site configuration specific to what you want, and then rely on the conversion process to turn that into the `modelconfigSDK.ModelConfiguration`. Which for what you are doing is just overhead/noise.

File: cmd/frontend/internal/publicrestapi/golly-recordings/chat_completions.recording.yaml (hunk 14 above)
Comment (chrsmith) [reply cmd/frontend/internal/publicrestapi/golly-recordings/chat_completions.recording.yaml]: That all makes sense. Just some random thoughts:

> I did it this way because I was trying to test assumptions like "every chat-capable model on dotcom should be able to go through this endpoint".

Yeah, we definitely need this integration test. But is _this_ the right place for it? Or some other more general `/tests/dotcom/all_models_actually_work_test.go`?

As far as just requiring that we re-record this test whenever we update `models.json`, that's fine too. I assume it's just a matter of running a couple of commands? If you touch this code in the near term, consider adding those lines to the `cmd/cody-gateway-config/README.md` file. That might save some trouble in the future when people unintentionally break these tests, because they don't know to regenerate the recordings here.

**Comment by chrsmith** [comment]
LGTM! Just a few minor notes and suggestions for cleaning things up in a subsequent PR. (e.g. adding the steps to regenerate the tests in `cmd/cody-gateway-config/README.md`, since that is now required.)

File: cmd/frontend/internal/publicrestapi/chat_completions_handler.go (hunk 10 above)
Comment (olafurpg) [reply cmd/frontend/internal/publicrestapi/chat_completions_handler.go:215]: I am planning to address the docs story separately, and I expect that will avoid these kinds of confusions in the future. It's too much effort to read through the Go comments to understand the purpose of our different APIs, we need higher-level docs that everybody can read (internal teammates + customers) explaining the motivation/context of different endpoints.

Comment (olafurpg) [reply cmd/frontend/internal/publicrestapi/chat_completions_handler.go:215]: The conceptual model I have in mind is that we are a bit like OpenRouter (https://openrouter.ai/). We have invested a ton in infrastructure to route LLM requests to different underlying LLM providers. OpenRouter uses the same API-contract as OpenAI (and we're adding here).

This is not about sending requests to OpenAI specifically, it's agnostic about the LLM provider.

File: cmd/frontend/internal/publicrestapi/chat_completions_models.go (hunk 12 above)
Comment (olafurpg) [reply cmd/frontend/internal/publicrestapi/chat_completions_models.go:20]: Captured your reply here https://linear.app/sourcegraph/issue/CODY-3222/consider-supporting-non-exact-model-parameter#comment-a622cffe

File: cmd/frontend/internal/publicrestapi/golly-recordings/chat_completions.recording.yaml (hunk 14 above)
Comment (olafurpg) [reply cmd/frontend/internal/publicrestapi/golly-recordings/chat_completions.recording.yaml]: We don't have to update this test whenever we update `models.json`. These new tests can use the recording for as long as we like. These tests are only stressing that the translation for `CodyCompletionRequestParameters` and `CompletionResponse` work as expected. Think of the recording as an alternative to hardcoding all the mocks with Go structs. The benefit of recordings is that they're easier to keep in sync with production data.

File: cmd/frontend/internal/publicrestapi/openai_handlers.go (hunk 1 above)
Comment (camdencheek) [reply cmd/frontend/internal/publicrestapi/openai_handlers.go]: FYI, claiming `/api/*` conflicts with `/api/console`, making the GraphQL console [currently inaccessible](https://linear.app/sourcegraph/issue/SRCH-898/issue-with-navigating-to-api-console-on).
